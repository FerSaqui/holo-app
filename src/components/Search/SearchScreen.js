import React, { useMemo } from 'react'
import { useLocation, useNavigate, useSearchParams } from 'react-router-dom';
import { getVtubersByName } from '../../helpers/getVtubersByName';
import { queryString } from '../../helpers/queryString';
import { useForm } from '../../hooks/useForm';
import { VtuberCard } from '../Vtuber/VtuberCard';

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();
  
  const { q } = queryString(location.search);

  const [formValues, handleInputChange, reset] = useForm({
    searchText: q
  });

  const { searchText } = formValues;

  const vtubersFilter = useMemo(() => getVtubersByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${ searchText }`); //Navegar en la misma ruta, es decir, ese string se mostrará en la url dentro de la misma página que estamos
  }

  return (
    <>
      <h1>Search</h1>
      <hr />

      <form onSubmit={handleSearch} className="container">

        <div className="row mb-3">
          <div className="col-9 col-lg-10">
            <input
              type="text"
              className="form-control"
              name="searchText"
              autoComplete="off"
              id="searchInput"
              value={searchText}
              onChange={handleInputChange}
            />
          </div>
          <div className="col-3 col-lg-2">
            <button type="submit" className="btn btn-primary w-100">Search</button>
          </div>
        </div>

        <div className="row">
          <h4>Results</h4>
          <hr />

          {
            (q === "")
                ? <div className="alert alert-info text-center">Search a vtuber!</div>
                : (vtubersFilter.length === 0) 
                    && <div className="alert alert-danger text-center">No results: {q}</div>
          }

          {
            vtubersFilter.map(vtuber => (
              <VtuberCard
                key={vtuber.id}
                {...vtuber}
              />
            ))
          }
        </div>

      </form>

    </>
  )
}