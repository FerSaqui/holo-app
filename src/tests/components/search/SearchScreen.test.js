import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/Search/SearchScreen";

/** Simular el llamado del hook useNavigate() de la librería react-router-dom
 * IMPORTANTE: debe iniciar con la palabra mock y no otro nombre
*/
const mockNavigate = jest.fn();

/**Sirve para evaluar cualquier hook de cualquier librería */
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockNavigate
}));
/*****************************************/

describe('Pruebas en <SearchScreen />', () => {

    test('Debe de mostrarse coreectamente con valores por defecto', () => {
        /**Es importante definir la ruta con initialEntries porque es lo que espera el MemoryRouter */
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find(".alert-info").text().trim() ).toBe("Search a vtuber!");

        
    });
    
    test('Debe de mostrar una vtuber y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=suisei"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper.find("input").prop("value") ).toBe("suisei");
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrar un mensaje de error si no encuentra una vtuber', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=suisei123"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper.find(".alert-danger").text().trim() ).toBe("No results: suisei123");
    });

    test('Debe de llamarse el navigate del hooks useNavigate() en el submit del form', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find("input").simulate("change", {
            target: {
                name: "searchText",
                value: "suisei"
            }
        });

        wrapper.find("form").prop("onSubmit")({
            preventDefault: () => {}
        });

        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalledWith("?q=suisei");
    });
});