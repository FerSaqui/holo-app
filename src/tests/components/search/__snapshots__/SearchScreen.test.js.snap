// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrar una vtuber y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <form
    className="container"
    onSubmit={[Function]}
  >
    <div
      className="row mb-3"
    >
      <div
        className="col-9 col-lg-10"
      >
        <input
          autoComplete="off"
          className="form-control"
          id="searchInput"
          name="searchText"
          onChange={[Function]}
          type="text"
          value="suisei"
        />
      </div>
      <div
        className="col-3 col-lg-2"
      >
        <button
          className="btn btn-primary w-100"
          type="submit"
        >
          Search
        </button>
      </div>
    </div>
    <div
      className="row"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col-12 col-md-6 col-lg-4 my-5 text-center animate__animated animate__fadeIn"
      >
        <div
          className="card text-white"
        >
          <img
            alt="Hoshimachi Suisei"
            className="card-img img-fluid"
            src="/assets/JP-Hoshimachi-Suisei.png"
          />
          <div
            className="card-img-overlay"
          >
            <a
              href="/vtuber/JP-Hoshimachi-Suisei"
              onClick={[Function]}
            >
              <h5
                className="card-title bg-info rounded-pill"
              >
                Hoshimachi Suisei
              </h5>
            </a>
            <p
              className="card-text bg-warning rounded-pill"
            >
              Hololive JP
            </p>
            <p
              className="card-text bg-info rounded position-absolute bottom-0 start-0"
            >
              On 12 November, she debuted her third 2D costume, in this costume she wears a black dress with a blue jacket, both the dress and the jacket are decorated with blue ribbons and laces, she wears a black belt with the words: 'Sphere' and 'Comet', she uses orange socks and black and short black boots
            </p>
          </div>
        </div>
      </div>
    </div>
  </form>,
]
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrarse coreectamente con valores por defecto 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <form
    className="container"
    onSubmit={[Function]}
  >
    <div
      className="row mb-3"
    >
      <div
        className="col-9 col-lg-10"
      >
        <input
          autoComplete="off"
          className="form-control"
          id="searchInput"
          name="searchText"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="col-3 col-lg-2"
      >
        <button
          className="btn btn-primary w-100"
          type="submit"
        >
          Search
        </button>
      </div>
    </div>
    <div
      className="row"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info text-center"
      >
        Search a vtuber!
      </div>
    </div>
  </form>,
]
`;
